docs/389ds/design/accesslog-json-design.md

```diff
@@ -0,0 +1,119 @@
+---
+title: "Access Log JSON Format Design"
+---
+
+# Access Log JSON Format
+----------------
+
+Overview
+--------
+
+In order to make the access logging more consumable by standard parsing tools there should be an option for writing the log in JSON
+
+JSON Design
+------------
+
+    {
+        // Header
+        version: "",
+        instance: "",
+        // The rest of the log
+        op: (
+            ABANDON, ADD, AUTOBIND, BIND/UNBIND, CLOSE, CMP, CONN, DEL, EXT,
+            MOD, MODRDN, SEARCH, SORT, STAT, RESULT, VLV
+        ),
+        local_time: <strftime output - customizable>,
+        gm_time: <gm time - uses a fixed format of **%FT%TZ**>,
+        conn_id: ####,
+        op_id: ####,
+        internal_op_id: ####,
+        internal_op_sub_id: ####,
+        client_ip: IP_ADDRESS/local,
+        server_ip: IP_ADDRESS//run/slapd-localhost.socket,
+        starttls: True/False,
+        authzid: DN,
+        internal: True/False
+        // BIND, UNBIND, AUTOBIND
+        bind_dn: DN,
+        method: "",
+        mech: "",
+        version: 2/3,
+        // ADD, CMP, MOD, DEL, MODRDN
+        target_dn: "", 
+        // CONN
+        fd: ####,
+        slot: ####,
+        tls: True/False,
+        // SEARCH
+        base: DN,
+        filter: "",
+        scope: ####,
+        attrs: [attr, attr,],
+        options: "",
+        // MODRDN
+        newsuperior: DN,
+        deleteoldrdn: True/False,
+        // VLV & SORT
+        sort: [attr, attr,],
+        vlv_index: "",
+        // ABANDON
+        msgid: ####,
+        targetop: "",
+        // EXT
+        oid: "",
+        name: "",
+        // STAT
+        stat_attr: "",
+        stat_key; "",
+        stat_count: ####,
+        stat_duration: ####
+        // RESULT
+        err: ####,
+        nentries: ####,
+        optime: ####,
+        wtime: ####,
+        etime: ####,
+        notes: "".
+        tag: ####,
+        details: "",
+        msg: "", // Needed?  Password expired message
+        pr_idx: ####,
+        pr_cookie: ####,
+        sid: "",
+        // Misc/future
+        ...,
+    },
+]
+```
+
+All event operations will also include the client & server IP addresses and the bind dn.
+
+Configuration
+------------------------
+
+Added a new configuration setting for access logging under **cn=config**
+
+```
+nsslapd-accesslog-json-format: default | json | json-pretty
```

Comment:
I wander if we should consider adding an option for selective JSON logging, allowing admin to choose which operations or fields are included in the JSON format (or exclude list?). But it might be too much to handle (and not very needed?). I don't have a hard opinion here.

Reply:
Do you mean filter the log?  Or a log mixed with JSON and non-JSON logging in it?

Personally I don't think we should filter access log content, and I don't think we should mix json with non-json logging.

Reply:
I mean filter the log entries before printing them as to include some fields or not. So log entries might look cleaner and consume less storage. 
But I'm also not sure if it worths it... Just an idea to consider here:)